consensus/tendermint/tendermint.go

```diff
@@ -0,0 +1,637 @@
+
+					Since the value's id is expected to be unique the id can be used to compare the values.
+
+				*/
+
+				if vr == nil && proposalFromProposer && t.state.step == propose {
+					vote := Prevote[H, A]{
+						Vote: Vote[H, A]{
+							Height: t.state.height,
+							Round:  t.state.round,
+							ID:     nil,
+							Sender: t.nodeAddr,
+						},
+					}
+
+					if validProposal && (t.state.lockedRound == nil || (*t.state.lockedValue).Hash() == vID) {
+						vote.ID = &vID
+					}
+
+					t.messages.addPrevote(vote)
+					t.broadcasters.PrevoteBroadcaster.Broadcast(vote)
+					t.state.step = prevote
+				}
+
+				/*
+					Check the upon condition on  line 28:
+
+						28: upon {PROPOSAL, h_p, round_p, v, vr} from proposer(h_p, round_p) AND 2f + 1 {PREVOTE,h_p, vr, id(v)} while
+							step_p = propose ∧ (vr ≥ 0 ∧ vr < round_p) do
+						29: if valid(v) ∧ (lockedRound_p ≤ vr ∨ lockedValue_p = v) then
+						30: 	broadcast {PREVOTE, hp, round_p, id(v)}
+						31: else
+						32:  	broadcast {PREVOTE, hp, round_p, nil}
+						33: step_p ← prevote
+
+				*/
+
+				if vr != nil && proposalFromProposer && t.state.step == propose {
+					_, prevotesForHVr, _ := t.messages.allMessages(p.Height, *vr)
+
+					var vals []A
+
+					for addr, valPrevotes := range prevotesForHVr {
+						for _, p := range valPrevotes {
+							if *p.ID == vID {
+								vals = append(vals, addr)
+							}
+						}
+					}
+
+					if t.validatorSetVotingPower(vals) >= q(t.validators.TotalVotingPower(p.Height)) && *vr >= uint(0) && *vr < t.state.round {
+						vote := Prevote[H, A]{
+							Vote: Vote[H, A]{
+								Height: t.state.height,
+								Round:  t.state.round,
+								ID:     nil,
+								Sender: t.nodeAddr,
+							},
+						}
+
+						if validProposal && (*t.state.lockedRound >= *vr || (*t.state.lockedValue).Hash() == vID) {
+							vote.ID = &vID
+						}
+
+						t.messages.addPrevote(vote)
+						t.broadcasters.PrevoteBroadcaster.Broadcast(vote)
+						t.state.step = prevote
+					}
+				}
+
+				/*
+					Check upon condition on line 36:
+
+						36: upon {PROPOSAL, h_p, round_p, v, ∗} from proposer(h_p, round_p) AND 2f + 1 {PREVOTE, h_p, round_p, id(v)} while
+							valid(v) ∧ step_p ≥ prevote for the first time do
+						37: if step_p = prevote then
+						38: 	lockedValue_p ← v
+						39: 	lockedRound_p ← round_p
+						40: 	broadcast {PRECOMMIT, h_p, round_p, id(v))}
+						41: 	step_p ← precommit
+						42: validValue_p ← v
+						43: validRound_p ← round_p
+				*/
+
+				if validProposal && t.state.step >= prevote {
+					// line 36: 36: upon hPROPOSAL, hp, roundp, v, ∗i from proposer(hp, roundp) AND 2f + 1 hPREVOTE, hp, roundp, id(v)i while
+					// valid(v) ∧ stepp ≥ prevote for the first time do
+					var vals []A
+
+					for addr, valPrevotes := range prevotesForHR {
+						for _, p := range valPrevotes {
+							if *p.ID == vID {
+								vals = append(vals, addr)
+							}
+						}
+					}
+
+				}
```

Comment:
What are these lines doing??

Reply:
These lines are fetching all the validators which prevote for `vID`, so that I check if there is a quorum.